lib/Parser/BGParseManager.cpp

```diff
@@ -0,0 +1,467 @@
+CriticalSection BGParseManager::s_staticMemberLock;
+
+// Static member management
+BGParseManager* BGParseManager::GetBGParseManager()
+{
+    AutoCriticalSection lock(&s_staticMemberLock);
+    if (s_BGParseManager == nullptr)
+    {        
+        AUTO_NESTED_HANDLED_EXCEPTION_TYPE(ExceptionType_DisableCheck);
+        s_BGParseManager = HeapNew(BGParseManager);
+        s_BGParseManager->Processor()->AddManager(s_BGParseManager);
+    }
+    return s_BGParseManager;
+}
+
+void BGParseManager::DeleteBGParseManager()
+{
+    AutoCriticalSection lock(&s_staticMemberLock);
+    if (s_BGParseManager != nullptr)
+    {
+        BGParseManager* manager = s_BGParseManager;
+        s_BGParseManager = nullptr;
+        HeapDelete(manager);
+    }
+}
+
+DWORD BGParseManager::GetNextCookie()
+{
+    AutoCriticalSection lock(&s_staticMemberLock);
+    return ++s_lastCookie;
+}
+
+
+// Note: runs on any thread
+BGParseManager::BGParseManager()
+    : JsUtil::WaitableJobManager(ThreadBoundThreadContextManager::GetSharedJobProcessor())
+{
+}
+
+// Note: runs on any thread
+BGParseManager::~BGParseManager()
+{
+    // First, remove the manager from the JobProcessor so that any remaining jobs
+    // are moved back into this manager's processed workitem list
+    if (Processor()->HasManager(this))
+    {
+        Processor()->RemoveManager(this);
+    }
+
+    Assert(this->workitemsProcessing.IsEmpty());
+
+    // Now, free all remaining processed jobs
+    BGParseWorkItem *item = (BGParseWorkItem*)this->workitemsProcessed.Head();
+    while (item != nullptr)
+    {
+        // Get the next item first so that the current item
+        // can be safely removed and freed
+        BGParseWorkItem* temp = item;
+        item = (BGParseWorkItem*)item->Next();
+
+        this->workitemsProcessed.Unlink(temp);
+        HeapDelete(temp);
+    }
+}
+
+// Returns the BGParseWorkItem that matches the provided cookie
+// Note: runs on any thread
+BGParseWorkItem* BGParseManager::FindJob(DWORD dwCookie, bool waitForResults)
+{
+    AutoOptionalCriticalSection autoLock(Processor()->GetCriticalSection());
+
+    Assert(dwCookie != 0);
+    BGParseWorkItem* matchedWorkitem = nullptr;
+
+    // First, look among processed jobs
+    for (BGParseWorkItem *item = (BGParseWorkItem*)this->workitemsProcessed.Head(); item != nullptr && matchedWorkitem == nullptr; item = (BGParseWorkItem*)item->Next())
+    {
+        if (item->GetCookie() == dwCookie)
+        {
+            matchedWorkitem = item;
+        }
+    }
+
+    if (matchedWorkitem == nullptr)
+    {
+        // Then, look among processing jobs
+        for (BGParseWorkItem *item = (BGParseWorkItem*)this->workitemsProcessing.Head(); item != nullptr && matchedWorkitem == nullptr; item = (BGParseWorkItem*)item->Next())
+        {
+            if (item->GetCookie() == dwCookie)
+            {
+                matchedWorkitem = item;
+            }
+        }
+
+        // Lastly, look among active jobs
```

Comment:
>active [](start = 30, length = 6)

nit: elsewhere, you used the word "queued" to describe this state, which I think is clearer than "active"